test/libsolutil/LP.cpp

```diff
@@ -0,0 +1,353 @@
+	feasible({{x, "0"}, {y, "1"}, {z, "2"}});
+	solver.pop();
+}
+
+BOOST_AUTO_TEST_CASE(splittable)
+{
+	Expression x = variable("x");
+	Expression y = variable("y");
+	Expression z = variable("z");
+	Expression w = variable("w");
+	solver.addAssertion(x < y);
+	solver.addAssertion(x < y - 2);
+	solver.addAssertion(z + w == 28);
+
+	solver.push();
+	solver.addAssertion(z >= 30);
+	infeasible();
+	solver.pop();
+
+	solver.addAssertion(z >= 2);
+	feasible({{x, "0"}, {y, "3"}, {z, "2"}, {w, "26"}});
+	solver.push();
+	solver.addAssertion(z >= 4);
+	feasible({{x, "0"}, {y, "3"}, {z, "4"}, {w, "24"}});
+
+	solver.push();
+	solver.addAssertion(z < 4);
+	infeasible();
+	solver.pop();
+
+	// z >= 4 is still active
+	solver.addAssertion(z >= 3);
+	feasible({{x, "0"}, {y, "3"}, {z, "4"}, {w, "24"}});
+}
+
+BOOST_AUTO_TEST_CASE(boolean)
+{
+	Expression x = variable("x");
+	Expression y = variable("y");
+	Expression z = variable("z");
+	solver.addAssertion(x <= 5);
+	solver.addAssertion(y <= 2);
+	solver.push();
+	solver.addAssertion(x < y && x > y);
+	infeasible();
+	solver.pop();
+	Expression w = booleanVariable("w");
+	solver.addAssertion(w == (x < y));
+	solver.addAssertion(w || x > y);
+	feasible({{x, "0"}, {y, "3"}, {z, "2"}, {w, "26"}});
+}
+
+BOOST_AUTO_TEST_CASE(boolean_complex)
+{
+	Expression x = variable("x");
+	Expression y = variable("y");
+	Expression a = booleanVariable("a");
+	Expression b = booleanVariable("b");
+	solver.addAssertion(x <= 5);
+	solver.addAssertion(y <= 2);
+	solver.addAssertion(a == (x >= 2));
+	solver.addAssertion(a || b);
+	solver.addAssertion(b == !a);
+	solver.addAssertion(b == (x < 2));
+	feasible({{a, "1"}, {b, "0"}, {x, "5"}, {y, "2"}});
+	solver.addAssertion(a && b);
+	infeasible();
+}
+
+BOOST_AUTO_TEST_CASE(magic_square)
+{
+	vector<Expression> vars;
+	for (size_t i = 0; i < 9; i++)
+		vars.push_back(variable(string{static_cast<char>('a' + i)}));
+	for (Expression const& var: vars)
+		solver.addAssertion(1 <= var && var <= 9);
+	// If we assert all to be mutually distinct, the problems gets too large.
+	for (size_t i = 0; i < 9; i++)
+		for (size_t j = i + 7; j < 9; j++)
+			solver.addAssertion(vars[i] != vars[j]);
+	for (size_t i = 0; i < 4; i++)
+		solver.addAssertion(vars[i] != vars[i + 1]);
+	for (size_t i = 0; i < 3; i++)
+		solver.addAssertion(vars[i] + vars[i + 3] + vars[i + 6] == 15);
+	for (size_t i = 0; i < 9; i += 3)
+		solver.addAssertion(vars[i] + vars[i + 1] + vars[i + 2] == 15);
+	feasible({
+		{vars[0], "1"}, {vars[1], "0"}, {vars[2], "5"},
+		{vars[3], "1"}, {vars[4], "0"}, {vars[5], "5"},
+		{vars[6], "1"}, {vars[7], "0"}, {vars[8], "5"}
+	});
+}
+
+BOOST_AUTO_TEST_CASE(boolean_complex_2)
+{
+	Expression x = variable("x");
+	Expression y = variable("y");
+	Expression a = booleanVariable("a");
+	Expression b = booleanVariable("b");
+	solver.addAssertion(x != 20);
+	feasible({{x, "19"}});
+	solver.addAssertion(x <= 5 || (x > 7 && x != 8));
+	solver.addAssertion(a = (x == 9));
```

Comment:
Is `a =` or `a ==` expected here?

With equality the assertions don't seem to hold, so this must be using assignment (?) or something is broken?

Reply:
Ok I think this may be just ignored: https://github.com/ethereum/solidity/pull/10714/files#r668988380